Number_types/include/CGAL/boost_mp.h

```diff
@@ -0,0 +1,537 @@
+: Coercion_traits <
+typename boost::multiprecision::detail::expression<T1,T2,T3,T4,T5>::result_type,
+boost::multiprecision::number<B, E> >
+{ };
+
+// TODO: coercion with expressions, fix existing coercions
+// (double -> rational is implicit only for 1.56+, see ticket #10082)
+// The real solution would be to avoid specializing Coercion_traits for all pairs of number types and let it auto-detect what works, so only broken types need an explicit specialization.
+
+// Ignore types smaller than long
+#define CGAL_COERCE_INT(int) \
+template <class B1, boost::multiprecision::expression_template_option E1> \
+struct Coercion_traits<boost::multiprecision::number<B1, E1>, int> { \
+  typedef boost::multiprecision::number<B1, E1> Type; \
+  typedef Tag_true Are_implicit_interoperable; \
+  typedef Tag_true Are_explicit_interoperable; \
+  struct Cast{ \
+    typedef Type result_type; \
+    template <class U> Type operator()(const U& x) const { return Type(x); } \
+  }; \
+}; \
+template <class B1, boost::multiprecision::expression_template_option E1> \
+struct Coercion_traits<int, boost::multiprecision::number<B1, E1> > \
+: Coercion_traits<boost::multiprecision::number<B1, E1>, int> {}
+
+CGAL_COERCE_INT(short);
+CGAL_COERCE_INT(int);
+CGAL_COERCE_INT(long);
+#undef CGAL_COERCE_INT
+
+// Ignore bounded-precision rationals
+#define CGAL_COERCE_FLOAT(float) \
+template <class B1, boost::multiprecision::expression_template_option E1> \
+struct Coercion_traits<boost::multiprecision::number<B1, E1>, float> { \
+  typedef boost::multiprecision::number<B1, E1> Type; \
+  typedef Boolean_tag<boost::multiprecision::number_category<Type>::value != boost::multiprecision::number_kind_integer> Are_implicit_interoperable; \
+  typedef Are_implicit_interoperable Are_explicit_interoperable; \
+  struct Cast{ \
+    typedef Type result_type; \
+    template <class U> Type operator()(const U& x) const { return Type(x); } \
+  }; \
+}; \
+template <class B1, boost::multiprecision::expression_template_option E1> \
+struct Coercion_traits<float, boost::multiprecision::number<B1, E1> > \
+: Coercion_traits<boost::multiprecision::number<B1, E1>, float> {}
+
+CGAL_COERCE_FLOAT(float);
+CGAL_COERCE_FLOAT(double);
+#undef CGAL_COERCE_FLOAT
+
+// Because of https://github.com/boostorg/multiprecision/issues/29 , this is not perfect and fails to read some KDS files.
+
+template <>
+class Input_rep<boost::multiprecision::cpp_rational> : public IO_rep_is_specialized {
+    boost::multiprecision::cpp_rational& q;
+public:
+    Input_rep(boost::multiprecision::cpp_rational& qq) : q(qq) {}
+    std::istream& operator()(std::istream& in) const {
+      internal::read_float_or_quotient<boost::multiprecision::cpp_int,boost::multiprecision::cpp_rational>(in, q);
+      return in;
+    }
+};
+#ifdef CGAL_USE_GMP
+template <>
+class Input_rep<boost::multiprecision::mpq_rational> : public IO_rep_is_specialized {
+    boost::multiprecision::mpq_rational& q;
```

Comment:
As far as I understand, they are not needed because `operator>>` is working fine. Input_rep is specialized because we are not happy with the default behavior of `operator<<`.

Reply:
Alright. Thanks for the explanation.